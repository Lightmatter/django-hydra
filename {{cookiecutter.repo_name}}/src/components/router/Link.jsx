/* eslint-disable jsx-a11y/anchor-has-content */
import React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import { useRouter } from 'next/router';
import NextLink from 'next/link';
import MuiLink from '@material-ui/core/Link';

const NextComposed = React.forwardRef(function NextComposed(props, ref) {
    const { as, href, ...other } = props;

    return (
        <NextLink href={href} as={as}>
            <a ref={ref} {...other} />
        </NextLink>
    );
});

NextComposed.propTypes = {
    as: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    href: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    prefetch: PropTypes.bool,
};

// A styled version of the Next.js Link component:
// https://nextjs.org/docs/#with-link
function Link(props) {
    const {
        href,
        activeClassName = 'active',
        className: classNameProps,
        innerRef,
        variant,
        naked,
        ...other
    } = props;
    const router = useRouter();
    const pathname = typeof href === 'string' ? href : href.pathname;
    const className = clsx(classNameProps, {
        [activeClassName]: router?.pathname === pathname && activeClassName,
    });

    if (naked) {
        return <NextComposed className={className} ref={innerRef} href={href} {...other} />;
    }

    return (
        <MuiLink
            component={NextComposed}
            className={className}
            ref={innerRef}
            href={href}
            variant={variant}
            {...other}
        />
    );
}

Link.propTypes = {
    activeClassName: PropTypes.string,
    as: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    className: PropTypes.string,
    href: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    innerRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    naked: PropTypes.bool,
    onClick: PropTypes.func,
    prefetch: PropTypes.bool,
    variant: PropTypes.string,
};

export default React.forwardRef((props, ref) => <Link {...props} innerRef={ref} />);
