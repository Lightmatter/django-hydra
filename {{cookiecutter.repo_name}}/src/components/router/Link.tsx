/* eslint-disable jsx-a11y/anchor-has-content, react/require-default-props */
import React from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import { useRouter } from 'next/router';
import NextLink from 'next/link';
import MuiLink from '@material-ui/core/Link';
import isServer from 'util/isServer';
import { Button } from '@material-ui/core';

const NextComposed = React.forwardRef(function NextComposed(
  { as, href, prefetch, ...other },
  ref
) {
  return (
    <NextLink href={href} as={as} prefetch={prefetch}>
      <a ref={ref} {...other} />
    </NextLink>
  );
});

NextComposed.propTypes = {
  as: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  href: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  prefetch: PropTypes.bool,
};

// A styled version of the Next.js Link component:
// https://nextjs.org/docs/#with-link
function Link({
  href,
  activeClassName = 'active',
  className: classNameProps,
  innerRef,
  variant,
  naked,
  componentType = 'link',
  ...other
}) {
  const router = useRouter();
  const pathname = typeof href === 'string' ? href : href.pathname;
  const className = clsx(classNameProps, {
    [activeClassName]: router?.pathname === pathname && activeClassName,
  });

  if (naked) {
    return (
      <NextComposed
        className={className}
        ref={innerRef}
        href={href}
        {...other}
      />
    );
  }

  const isExternalLink = linkHref => {
    const checkHref = `${window.location.protocol}//${window.location.host}`;
    return !(
      !linkHref ||
      linkHref[0] === '?' ||
      linkHref[0] === '/' ||
      linkHref[0] === '#' ||
      linkHref.startsWith(checkHref) ||
      linkHref.startsWith(window.location.host) ||
      linkHref.substring(0, 4) === 'tel:' ||
      linkHref.substring(0, 7) === 'mailto:'
    );
  };

  let target;
  if (!isServer()) {
    target = isExternalLink(href) ? '_blank' : undefined;
  }
  let LinkComponent = MuiLink;
  if (componentType === 'button') {
    LinkComponent = Button;
  }
  return (
    <LinkComponent
      target={target}
      component={NextComposed}
      className={className}
      ref={innerRef}
      href={href}
      variant={variant}
      {...other}
    />
  );
}

Link.propTypes = {
  activeClassName: PropTypes.string,
  as: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  className: PropTypes.string,
  href: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  innerRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
  naked: PropTypes.bool,
  onClick: PropTypes.func,
  prefetch: PropTypes.bool,
  variant: PropTypes.string,
  componentType: PropTypes.oneOf(['button', 'link']),
};

export default React.forwardRef((props, ref) => (
  <Link {...props} innerRef={ref} />
));
