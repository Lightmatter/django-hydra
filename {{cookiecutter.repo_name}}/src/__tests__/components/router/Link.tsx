import React from 'react';
//import { createMount } from '@material-ui/core/test-utils';

import renderer from 'react-test-renderer';
import Link from 'components/router/Link';
import { shallow } from 'enzyme';

describe('Link', () => {
  it('renders Link unchanged', () => {
    const link = mount(<Link href="/signup">link</Link>);
    expect(link).toMatchSnapshot();
  });
  it('opens offsite links in a new tab', () => {
    const link = shallow(<Link href="http://www.google.com">link</Link>);
    expect(link.html()).toMatch('target="_blank"');
  });
  it('doesnt open onsite links in a new tab', () => {
    const link = shallow(<Link href="/home">link</Link>);
    expect(link.html()).not.toMatch('target="_blank"');
  });
  it('attaches the active class to links on the current url', () => {});
});
