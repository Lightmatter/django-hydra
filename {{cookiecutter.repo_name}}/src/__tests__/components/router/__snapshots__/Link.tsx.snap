// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders Link unchanged 1`] = `
<ForwardRef
  href="/signup"
>
  <Link
    href="/signup"
    innerRef={null}
  >
    <WithStyles(ForwardRef(Link))
      className=""
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "as": [Function],
            "href": [Function],
            "prefetch": [Function],
          },
          "render": [Function],
        }
      }
      href="/signup"
    >
      <ForwardRef(Link)
        className=""
        classes={
          Object {
            "button": "MuiLink-button",
            "focusVisible": "Mui-focusVisible",
            "root": "MuiLink-root",
            "underlineAlways": "MuiLink-underlineAlways",
            "underlineHover": "MuiLink-underlineHover",
            "underlineNone": "MuiLink-underlineNone",
          }
        }
        component={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "propTypes": Object {
              "as": [Function],
              "href": [Function],
              "prefetch": [Function],
            },
            "render": [Function],
          }
        }
        href="/signup"
      >
        <WithStyles(ForwardRef(Typography))
          className="MuiLink-root MuiLink-underlineHover"
          color="primary"
          component={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": Object {
                "as": [Function],
                "href": [Function],
                "prefetch": [Function],
              },
              "render": [Function],
            }
          }
          href="/signup"
          onBlur={[Function]}
          onFocus={[Function]}
          variant="inherit"
        >
          <ForwardRef(Typography)
            className="MuiLink-root MuiLink-underlineHover"
            classes={
              Object {
                "alignCenter": "MuiTypography-alignCenter",
                "alignJustify": "MuiTypography-alignJustify",
                "alignLeft": "MuiTypography-alignLeft",
                "alignRight": "MuiTypography-alignRight",
                "body1": "MuiTypography-body1",
                "body2": "MuiTypography-body2",
                "button": "MuiTypography-button",
                "caption": "MuiTypography-caption",
                "colorError": "MuiTypography-colorError",
                "colorInherit": "MuiTypography-colorInherit",
                "colorPrimary": "MuiTypography-colorPrimary",
                "colorSecondary": "MuiTypography-colorSecondary",
                "colorTextPrimary": "MuiTypography-colorTextPrimary",
                "colorTextSecondary": "MuiTypography-colorTextSecondary",
                "displayBlock": "MuiTypography-displayBlock",
                "displayInline": "MuiTypography-displayInline",
                "gutterBottom": "MuiTypography-gutterBottom",
                "h1": "MuiTypography-h1",
                "h2": "MuiTypography-h2",
                "h3": "MuiTypography-h3",
                "h4": "MuiTypography-h4",
                "h5": "MuiTypography-h5",
                "h6": "MuiTypography-h6",
                "noWrap": "MuiTypography-noWrap",
                "overline": "MuiTypography-overline",
                "paragraph": "MuiTypography-paragraph",
                "root": "MuiTypography-root",
                "srOnly": "MuiTypography-srOnly",
                "subtitle1": "MuiTypography-subtitle1",
                "subtitle2": "MuiTypography-subtitle2",
              }
            }
            color="primary"
            component={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "propTypes": Object {
                  "as": [Function],
                  "href": [Function],
                  "prefetch": [Function],
                },
                "render": [Function],
              }
            }
            href="/signup"
            onBlur={[Function]}
            onFocus={[Function]}
            variant="inherit"
          >
            <ForwardRef(NextComposed)
              className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
              href="/signup"
              onBlur={[Function]}
              onFocus={[Function]}
            >
              <Link
                href="/signup"
              >
                <a
                  className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                  href="/signup"
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onMouseEnter={[Function]}
                >
                  link
                </a>
              </Link>
            </ForwardRef(NextComposed)>
          </ForwardRef(Typography)>
        </WithStyles(ForwardRef(Typography))>
      </ForwardRef(Link)>
    </WithStyles(ForwardRef(Link))>
  </Link>
</ForwardRef>
`;
